### DISCUSSION TITLE
Finite fields of order p^n where p is a prime.

### DISCUSSION THREAD

**Discussion by shamEiNew** [sympy/sympy#23540] [Ideas]
Hi all,

I went through Sympy documentation for finite fields there it is mentioned "It would be good to have a proper implementation of prime power fields (GF(p**n)) but these are not yet implemented in SymPY." Is anybody working on this implementation?
I tried defining like this but I got an error on sympy live shell
```
>>> K = FiniteExtension(Poly(x**3, x, modulus=16))
>>> K
GF(16)[x]/(x∗∗3)

>>> a = Poly(x, y, domain=K)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/base/data/home/apps/s~sympy-live-hrd/66.426491309333028408/sympy/sympy/polys/polytools.py", line 109, in __new__
    opt = options.build_options(gens, args)
  File "/base/data/home/apps/s~sympy-live-hrd/66.426491309333028408/sympy/sympy/polys/polyoptions.py", line 735, in build_options
    return Options(gens, args)
  File "/base/data/home/apps/s~sympy-live-hrd/66.426491309333028408/sympy/sympy/polys/polyoptions.py", line 154, in __init__
    preprocess_options(args)
  File "/base/data/home/apps/s~sympy-live-hrd/66.426491309333028408/sympy/sympy/polys/polyoptions.py", line 152, in preprocess_options
    self[option] = cls.preprocess(value)
  File "/base/data/home/apps/s~sympy-live-hrd/66.426491309333028408/sympy/sympy/polys/polyoptions.py", line 487, in preprocess
    raise OptionError('expected a valid domain specification, got %s' % domain)
OptionError: expected a valid domain specification, got GF(16)[x]/(x**3)
```
Note that above is not working for any modulo in sympy shell.

**Comment by jksuom** [comment]
See #13886 and #18693. It seems that nobody is working on this right now.

**Reply by shamEiNew** [reply]
thanks!
I'll go through the links and let you know if can do something

**Comment by skieffer** [comment]
This summary: https://github.com/sympy/sympy/issues/9544#issuecomment-647726095 seems especially helpful.

Was a decision ever made between approaches (3a) and (3b) described there?

**Reply by oscarbenjamin** [reply]
I don't think any decision was made. Probably I would go with `3a` though. There exist special methods and algorithms for `GF(p)[x]` in `galoistools` and those should be used for implementing `GF(p)[x]/q(x)` which is how `GF(p^n)` would be implemented under the hood.

The point of `3b` is that it is already possible to use `FiniteExtension` to get limited support for `GF(p^n)` e.g.:
```python
In [7]: from sympy.polys.agca.extensions import FiniteExtension

In [8]: x = symbols('x')

In [9]: F = FiniteExtension(Poly(x**3 - x + 1, x, modulus=3))

In [10]: xgf = F.from_sympy(x)

In [11]: xgf
Out[11]: x

In [12]: xgf**100
Out[12]: -x - 1
```
It would be nice if generic "extension" code could work equally well for arbitrary extensions as for specific cases like finite fields. I expect though that it would make things tricky because there would end up being some methods that you would want for finite fields that wouldn't be generally applicable in the case of arbitrary extensions. Of course we could make it so that FiniteExtension and any corresponding domain for FiniteFields have a common interface to the extent that it makes sense so that at least some algorithms could operate equally on either.

Also there is something of a cyclic level problem with FiniteExtension since it uses Poly whereas I think that Poly should be something that sits on top of the domain system. Something like `GF(p^n)` should really be implemented efficiently at a very low level.

**Reply by skieffer** [reply]
> Something like `GF(p^n)` should really be implemented efficiently at a very low level.

Yeah, that would be my sense as well.
